File: core/model/docker.go
package model

import "time"

// docker

type LogOption struct {
	LogMaxSize string `json:"log_max_size"`
	LogMaxFile string `json:"log_max_file"`
}

type Ipv6Option struct {
	FixedCidrV6  string `json:"fixed_cidr_v6"`
	Ip6Tables    bool   `json:"ip6_tables" validate:"required"`
	Experimental bool   `json:"experimental"`
}

type DaemonJsonUpdateByFile struct {
	File string `json:"file"`
}

type DaemonJsonConf struct {
	IsSwarm      bool     `json:"is_swarm"`
	Status       string   `json:"status"`
	Version      string   `json:"version"`
	Mirrors      []string `json:"registry_mirrors"`
	Registries   []string `json:"insecure_registries"`
	LiveRestore  bool     `json:"live_restore"`
	IPTables     bool     `json:"ip_tables"`
	CgroupDriver string   `json:"cgroup_driver"`

	Ipv6         bool   `json:"ipv6"`
	FixedCidrV6  string `json:"fixed_cidr_v6"`
	Ip6Tables    bool   `json:"ip6_tables"`
	Experimental bool   `json:"experimental"`

	LogMaxSize string `json:"log_max_size"`
	LogMaxFile string `json:"log_max_file"`
}

type DockerOperation struct {
	Operation string `json:"operation" validate:"required,oneof=start restart stop"`
}

// common
type Inspect struct {
	ID   string `json:"id" validate:"required"`
	Type string `json:"type" validate:"required,oneof=container image volume network"`
}

type Prune struct {
	PruneType  string `json:"type" validate:"required,oneof=container image volume network buildcache"`
	WithTagAll bool   `json:"with_tag_all"`
}

type PruneResult struct {
	DeletedNumber  int `json:"deleted_number"`
	SpaceReclaimed int `json:"space_reclaimed"`
}

// container
type QueryContainer struct {
	PageInfo
	Name    string `json:"name"`
	State   string `json:"state" validate:"required,oneof=all created running paused restarting removing exited dead"`
	OrderBy string `json:"order_by"`
	Order   string `json:"order"`
}

type ContainerInfo struct {
	ContainerID string `json:"container_id"`
	Name        string `json:"name"`
	ImageId     string `json:"image_id"`
	ImageName   string `json:"image_name"`
	CreateTime  string `json:"create_time"`
	State       string `json:"state"`
	RunTime     string `json:"run_time"`

	Network []string `json:"network"`
	Ports   []string `json:"ports"`

	From string `json:"from"`

	// AppName        string   `json:"app_name"`
	// AppInstallName string   `json:"app_install_name"`
	// Websites       []string `json:"websites"`
}

type VolumeHelper struct {
	Type         string `json:"type"`
	SourceDir    string `json:"source_dir"`
	ContainerDir string `json:"container_dir"`
	Mode         string `json:"mode"`
}
type PortHelper struct {
	HostIP        string `json:"host_ip"`
	HostPort      string `json:"host_port"`
	ContainerPort string `json:"container_port"`
	Protocol      string `json:"protocol"`
}

type ContainerOperate struct {
	ContainerID     string         `json:"container_id"`
	ForcePull       bool           `json:"force_pull"`
	Name            string         `json:"name" validate:"required"`
	Image           string         `json:"image" validate:"required"`
	Network         string         `json:"network"`
	Ipv4            string         `json:"ipv4"`
	Ipv6            string         `json:"ipv6"`
	PublishAllPorts bool           `json:"publish_all_ports"`
	ExposedPorts    []PortHelper   `json:"exposed_ports"`
	Tty             bool           `json:"tty"`
	OpenStdin       bool           `json:"open_stdin"`
	Cmd             []string       `json:"cmd"`
	Entrypoint      []string       `json:"entry_point"`
	CPUShares       int64          `json:"cpu_shares"`
	NanoCPUs        float64        `json:"nano_cpus"`
	Memory          float64        `json:"memory"`
	Privileged      bool           `json:"privileged"`
	AutoRemove      bool           `json:"auto_remove"`
	Volumes         []VolumeHelper `json:"volumes"`
	Labels          []string       `json:"labels"`
	Env             []string       `json:"env"`
	RestartPolicy   string         `json:"restart_policy"`
}

type ContainerUpgrade struct {
	Name      string `json:"name" validate:"required"`
	Image     string `json:"image" validate:"required"`
	ForcePull bool   `json:"force_pull"`
}

type ContainerOperation struct {
	Names     []string `json:"names" validate:"required"`
	Operation string   `json:"operation" validate:"required,oneof=start stop restart kill pause resume remove"`
}

type ContainerResourceUsage struct {
	ContainerID string `json:"container_id"`

	CPUTotalUsage uint64  `json:"cpu_total_usage"`
	SystemUsage   uint64  `json:"system_usage"`
	CPUPercent    float64 `json:"cpu_percent"`
	PercpuUsage   int     `json:"per_cpu_usage"`

	MemoryCache   uint64  `json:"memory_cache"`
	MemoryUsage   uint64  `json:"memory_usage"`
	MemoryLimit   uint64  `json:"memory_limit"`
	MemoryPercent float64 `json:"memory_percent"`
}

type ContainerStats struct {
	CPUPercent float64 `json:"cpu_percent"`
	Memory     float64 `json:"memory"`
	Cache      float64 `json:"cache"`
	IORead     float64 `json:"io_read"`
	IOWrite    float64 `json:"io_write"`
	NetworkRX  float64 `json:"network_rx"`
	NetworkTX  float64 `json:"network_tx"`

	ShotTime time.Time `json:"shot_time"`
}

type ContainerResourceLimit struct {
	CPU    int    `json:"cpu"`
	Memory uint64 `json:"memory"`
}

// compose
type ComposeInfo struct {
	Name            string             `json:"name"`
	CreatedAt       string             `json:"created_at"`
	CreatedBy       string             `json:"created_by"`
	ContainerNumber int                `json:"container_number"`
	ConfigFile      string             `json:"config_file"`
	Workdir         string             `json:"work_dir"`
	Path            string             `json:"path"`
	Containers      []ComposeContainer `json:"containers"`
}

type ComposeContainer struct {
	ContainerID string `json:"container_id"`
	Name        string `json:"name"`
	CreateTime  string `json:"create_time"`
	State       string `json:"state"`
}

type ComposeCreate struct {
	Name     string `json:"name"`
	From     string `json:"from" validate:"required,oneof=edit path template"`
	File     string `json:"file"`
	Path     string `json:"path"`
	Template uint   `json:"template"`
}

type ComposeOperation struct {
	Name      string `json:"name" validate:"required"`
	Path      string `json:"path" validate:"required"`
	Operation string `json:"operation" validate:"required,oneof=start stop down"`
	WithFile  bool   `json:"with_file"`
}

type ComposeUpdate struct {
	Name    string `json:"name" validate:"required"`
	Path    string `json:"path" validate:"required"`
	Content string `json:"content" validate:"required"`
}

// image
type Image struct {
	ID        string    `json:"id"`
	CreatedAt time.Time `json:"created_at"`
	IsUsed    bool      `json:"is_used"`
	Tags      []string  `json:"tags"`
	Size      string    `json:"size"`
}

type ImageLoad struct {
	Path string `json:"path" validate:"required"`
}

type ImageBuild struct {
	From       string   `json:"from" validate:"required"`
	Name       string   `json:"name" validate:"required"`
	Dockerfile string   `json:"docker_file" validate:"required"`
	Tags       []string `json:"tags"`
}

type ImagePull struct {
	ImageName string `json:"image_name" validate:"required"`
}

type ImageTag struct {
	SourceID   string `json:"source_id" validate:"required"`
	TargetName string `json:"target_name" validate:"required"`
}

type ImagePush struct {
	TagName string `json:"tag_name" validate:"required"`
	Name    string `json:"name" validate:"required"`
}

type ImageSave struct {
	TagName string `json:"tag_name" validate:"required"`
	Path    string `json:"path" validate:"required"`
	Name    string `json:"name" validate:"required"`
}

// volume
type Volume struct {
	Name       string    `json:"name"`
	Labels     []string  `json:"labels"`
	Driver     string    `json:"driver"`
	Mountpoint string    `json:"mount_point"`
	CreatedAt  time.Time `json:"created_at"`
}

type VolumeCreate struct {
	Name    string   `json:"name" validate:"required"`
	Driver  string   `json:"driver" validate:"required"`
	Options []string `json:"options"`
	Labels  []string `json:"labels"`
}

// network
type Network struct {
	ID         string    `json:"id"`
	Name       string    `json:"name"`
	Labels     []string  `json:"labels"`
	Driver     string    `json:"driver"`
	IPAMDriver string    `json:"ipam_driver"`
	Subnet     string    `json:"subnet"`
	Gateway    string    `json:"gateway"`
	CreatedAt  time.Time `json:"created_at"`
	Attachable bool      `json:"attachable"`
}

type NetworkCreate struct {
	Name       string     `json:"name" validate:"required"`
	Driver     string     `json:"driver" validate:"required"`
	Options    []string   `json:"options"`
	Ipv4       bool       `json:"ipv4"`
	Subnet     string     `json:"subnet"`
	Gateway    string     `json:"gateway"`
	IPRange    string     `json:"ip_range"`
	AuxAddress []KeyValue `json:"aux_address"`

	Ipv6         bool       `json:"ipv6"`
	SubnetV6     string     `json:"subnet_v6"`
	GatewayV6    string     `json:"gateway_v6"`
	IPRangeV6    string     `json:"ip_range_v6"`
	AuxAddressV6 []KeyValue `json:"aux_address_v6"`
	Labels       []string   `json:"labels"`
}

File: frontend/src/api/docker.ts
import request from '@/helper/api-helper';
import { ApiListResult } from '@/types/global';

export interface ComposeCreateResult {
  log?: string;
}

export interface ComposeTestResult {
  error?: string;
  success?: boolean;
}

export interface PortHelper {
  container_port?: string;
  host_ip?: string;
  host_port?: string;
  protocol?: string;
}

export interface VolumeHelper {
  container_dir?: string;
  mode?: string;
  source_dir?: string;
  type?: string;
}

export interface ContainerOperate {
  auto_remove?: boolean;
  cmd?: string[];
  container_id?: string;
  cpu_shares?: number;
  entry_point?: string[];
  env?: string[];
  exposed_ports?: PortHelper[];
  force_pull?: boolean;
  image: string;
  ipv4?: string;
  ipv6?: string;
  labels?: string[];
  memory?: number;
  name: string;
  nano_cpus?: number;
  network?: string;
  open_stdin?: boolean;
  privileged?: boolean;
  publish_all_ports?: boolean;
  restart_policy?: string;
  tty?: boolean;
  volumes?: VolumeHelper[];
}

export interface ContainerOperation {
  names: string[];
  operation:
    | 'start'
    | 'stop'
    | 'restart'
    | 'kill'
    | 'pause'
    | 'unpause'
    | 'remove';
}

export interface ContainerResourceLimit {
  cpu?: number;
  memory?: number;
}

export interface ContainerStats {
  cache?: number;
  cpu_percent?: number;
  io_read?: number;
  io_write?: number;
  memory?: number;
  network_rx?: number;
  network_tx?: number;
  shot_time?: string;
}

export interface ContainerUpgrade {
  force_pull?: boolean;
  image: string;
  name: string;
}

export interface CreateCompose {
  compose_content?: string;
  env_content?: string;
  name?: string;
}

export interface DaemonJsonConf {
  cgroup_driver: string;
  experimental: boolean;
  fixed_cidr_v6: string;
  insecure_registries: string[];
  ip6_tables: boolean;
  ip_tables: boolean;
  ipv6: boolean;
  is_swarm: boolean;
  live_restore: boolean;
  log_max_file: string;
  log_max_size: string;
  registry_mirrors: string[];
  status: string;
  version: string;
}

export interface DockerOperation {
  operation: 'start' | 'restart' | 'stop';
}

export interface DockerStatus {
  status?: string;
}

export interface ImageBuild {
  docker_file: string;
  from: string; // edit|file
  name: string;
  tags?: string[];
}

export interface ImageLoad {
  path: string;
}

export interface ImagePull {
  image_name: string;
}

export interface ImagePush {
  name: string;
  tag_name: string;
}

export interface ImageSave {
  name: string;
  path: string;
  tag_name: string;
}

export interface ImageTag {
  source_id: string;
  target_name: string;
}

export interface Ipv6Option {
  experimental?: boolean;
  fixed_cidr_v6?: string;
  ip6_tables: boolean;
}

export interface KeyValue {
  key: string;
  value: string;
}

export interface LogOption {
  log_max_file?: string;
  log_max_size?: string;
}

export interface NetworkCreate {
  aux_address?: KeyValue[];
  aux_address_v6?: KeyValue[];
  driver: string;
  gateway?: string;
  gateway_v6?: string;
  ip_range?: string;
  ip_range_v6?: string;
  ipv4?: boolean;
  ipv6?: boolean;
  labels?: string[];
  name: string;
  options?: string[];
  subnet?: string;
  subnet_v6?: string;
}

export interface Prune {
  type: 'container' | 'image' | 'volume' | 'network' | 'buildcache';
  with_tag_all?: boolean;
}

export interface PruneResult {
  deleted_number?: number;
  space_reclaimed?: number;
}

export interface Rename {
  name: string;
  new_name: string;
}

export interface VolumeCreate {
  driver: string;
  labels?: string[];
  name: string;
  options?: string[];
}

// 查询 compose
export const queryComposeApi = (params: {
  info?: string;
  page: number;
  page_size: number;
}) => request.get<ApiListResult<any>>('/docker/{host}/compose', params);

// 查询 compose详情
export const getComposeDetailApi = (params: { name: string }) =>
  request.get('/docker/{host}/compose/detail', params);

// 更新 compose
export const updateComposeApi = (params: CreateCompose) =>
  request.put('/docker/{host}/compose', params);

// 创建 compose
export const createComposeApi = (params: CreateCompose) =>
  request.post<ComposeCreateResult>('/docker/{host}/compose', params);

// 操作 compose
export const operateComposeApi = (params: {
  name: string;
  operation: 'start' | 'stop' | 'restart' | 'up' | 'down';
}) => request.post('/docker/{host}/compose/operation', params);

// 删除 compose
export const deleteComposeApi = (params: { name: string }) =>
  request.delete('/docker/{host}/compose', params);

// 测试 compose
export const testComposeApi = (params: CreateCompose) =>
  request.post<ComposeTestResult>('/docker/{host}/compose/test', params);

// 获取 docker 配置
export const getDockerConfApi = () =>
  request.get<DaemonJsonConf>('/docker/{host}/conf');

// 更新 docker 配置 (修改字段)
export const updateDockerConfApi = (params: { key: string; value: any }) =>
  request.put('/docker/{host}/conf', params);

// 获取 docker 配置文件
export const getDockerConfRawApi = () =>
  request.get<{
    content: string;
  }>('/docker/{host}/conf/raw');

// 更新docker配置 (通过配置文件修改)
export const updateDockerConfRawApi = (params: { content: string }) =>
  request.put('/docker/{host}/conf/raw', params);

// 更新 ipv6 选项
export const updateIpv6OptionApi = (params: Ipv6Option) =>
  request.put('/docker/{host}/ipv6', params);

// 更新日志选项
export const updateLogOptionApi = (params: LogOption) =>
  request.put('/docker/{host}/log', params);

// 查询容器
export const queryContainersApi = (params: {
  info?: string;
  state: string;
  page: number;
  page_size: number;
  order_by?: string;
}) => request.get<ApiListResult<any>>('/docker/{host}/containers', params);

// 更新容器
export const updateContainerApi = (params: ContainerOperate) =>
  request.put('/docker/{host}/containers', params);

// 创建容器
export const createContainerApi = (params: ContainerOperate) =>
  request.post('/docker/{host}/containers', params);

// 获取容器详情
export const getContainerDetailApi = (id: number) =>
  request.get<ContainerOperate>('/docker/{host}/containers/detail', {
    id,
  });

// 获取容器资源限制
export const getContainerResourceLimitApi = () =>
  request.get<ContainerResourceLimit>('/docker/{host}/containers/limit');

// 清理容器日志
export const cleanContainerLogApi = (id: number) =>
  request.delete('/docker/{host}/containers/logs', { id });

// 查询容器名称
export const queryContainerNamesApi = () =>
  request.get<ApiListResult<any>>('/docker/{host}/containers/names');

// 退出容器终端
export const quitContainerTerminalApi = (params: {
  session: string;
  data?: string;
  type: 'screen';
}) => request.post('/docker/{host}/containers/terminal/quit', params);

// 容器批量操作
export const operateContainersApi = (params: ContainerOperation) =>
  request.post('/docker/{host}/containers/operation', params);

// 容器重命名
export const renameContainerApi = (params: Rename) =>
  request.post('/docker/{host}/containers/rename', params);

// 获取容器统计
export const getContainerStatsApi = (id: number) =>
  request.get<ContainerStats>('/docker/{host}/containers/stats', { id });

// 升级容器
export const upgradeContainerApi = (params: ContainerUpgrade) =>
  request.post('/docker/{host}/containers/upgrade', params);

// 获取容器资源使用列表
export const getContainerUsagesApi = () =>
  request.get<ApiListResult<any>>('/docker/{host}/containers/usages');

// 获取镜像
export const queryImagesApi = (params: {
  info?: string;
  page: number;
  page_size: number;
}) => request.get<ApiListResult<any>>('/docker/{host}/images', { ...params });

// 批量删除镜像
export const batchDeleteImagesApi = (params: {
  force: boolean;
  sources: string;
}) => request.delete('/docker/{host}/images', params);

// 构建镜像
export const buildImageApi = (params: ImageBuild) =>
  request.post('/docker/{host}/images/build', params);

// 导出镜像
export const exportImageApi = (params: ImageSave) =>
  request.post('/docker/{host}/images/export', params);

// 导入镜像
export const importImageApi = (params: ImageLoad) =>
  request.post('/docker/{host}/images/import', params);

// 查询镜像名称
export const queryImageNamesApi = () =>
  request.get<ApiListResult<any>>('/docker/{host}/images/names');

// 拉取镜像
export const pullImageApi = (params: ImagePull) =>
  request.post('/docker/{host}/images/pull', params);

// 推送镜像
export const pushImageApi = (id: string, params: ImagePush) =>
  request.post('/docker/{host}/images/push', params, {
    params: { id },
  });

// 设置镜像标签
export const setImageTagApi = (params: ImageTag) =>
  request.put('/docker/{host}/images/tag', params);

// inspect
export const inspectApi = (params: { type: string; id: string }) =>
  request.get<any>('/docker/{host}/inspect', { ...params });

// 获取网络
export const getNetworksApi = (params: {
  info?: string;
  page: number;
  page_size: number;
}) => request.get<ApiListResult<any>>('/docker/{host}/networks', params);

// 创建网络
export const createNetworkApi = (params: NetworkCreate) =>
  request.post('/docker/{host}/networks', params);

// 批量删除网络
export const batchDeleteNetworkApi = (params: {
  force: string;
  sources: string;
}) => request.delete('/docker/{host}/networks', params);

// 查询网络名称
export const queryNetworkNamesApi = () =>
  request.get<ApiListResult<any>>('/docker/{host}/networks/names');

// docker 操作
export const dockerOperationApi = (params: DockerOperation) =>
  request.post('/docker/{host}/operation', params);

// prune
export const pruneApi = (params: Prune) =>
  request.post<PruneResult>('/docker/{host}/prune', params);

// 获取 docker 状态
export const getDockerStatusApi = () =>
  request.get<DockerStatus>('/docker/{host}/status');

// 获取卷
export const getVolumesApi = (params: {
  info?: string;
  page: number;
  page_size: number;
}) => request.get<ApiListResult<any>>('/docker/{host}/volumes', params);

// 创建卷
export const createVolumeApi = (params: VolumeCreate) =>
  request.post('/docker/{host}/volumes', params);

// 批量删除卷
export const batchDeleteVolumeApi = (params: {
  force: string;
  sources: string;
}) => request.delete('/docker/{host}/volumes', params);

// 查询卷名称
export const queryVolumeNamesApi = () =>
  request.get<ApiListResult<any>>('/docker/{host}/volumes/names');
